docs/architecture/adr-073-libp2p.md

```diff
@@ -0,0 +1,189 @@
+code makes these changes relatively surgical, and limited to a small
+number of methods:
+
+- `p2p.Router.OpenChannel` will still return a `Channel` structure
+  which will continue to serve as a pipe between the reactors and the
+  `Router`. The implementation will no longer need the queue
+  implementation and this operation will instead start goroutines that
+  are responsible for routing the messages from the channel to libp2p
+  fundamentals, replacing the current `p2p.Router.routeChannel`.
+
+- The current `p2p.Rotuer.dialPeers` and `p2p.Router.acceptPeers`,
+  which are responsible for establishing outbound and inbound
+  connections respectively, will be removed, along with
+  `p2p.Router.openConnection` and the libp2p connection manager will
+  be responsible for maintaining network connectivity.
+
+- The `p2p.Channel` interface should change to replace the golang
+  channels with a more functional interface for sending messages. The
+  new methods on this object, will take contexts to support safe
+  cancellation and return errors, and generally block rather than
+  return asynchronously. The `Out` channel through which
+  reactors send messages to Peers, will be replaced by a `Send`
+  method, and the Error channel will be replaced by an `Error`
+  method. 
+
+- Rectors will have access to an interface that will allow them to
+  access information from the current state of the Peer information
+  from libp2p. This will supplant the  `p2p.PeerUpdates`
+  subscription.
+
+### Upgrade and Compatibility
+
+Because the routers and all current P2P code is in the `internal`
+package and not part of the public API, the only changes to the public
+API surface area of the tendermint will be different configuration
+file options, replacing the current P2P options with options relevant
+to libp2p.
+
+However, it will not be possible to run a network with both networking
+stacks active at once, so the upgrade to the version of Tendermint
+will need to be coordinated between all nodes of the network.
+
+## Open Questions
+
+- What is the role of Protocol Labs in the implementation of libp2p in
+  tendermint, both during the initial implementation and on an ongoing
+  basis thereafter?
+
+- Should all P2P traffic for a given node be pushed to a single topic,
+  and we assume that a topic maps to a specific ChainID, or should
+  each reactor (or type of message) have its own topic?
+
+- Tendermint presently has a concept of message-type priority, which
+  provides a very course QoS-like functionality and
+  intuitively/theoretically ensures that evidence and consensus
+  messages don't get starved by blocksync/statesync messages. It's
+  unclear if we should attempt to replicate this with libp2p and even
+  if we should.
+
+- Is it possible to attach additional (and potentially arbitrary)
+  information into the DHT as part of the heartbeats between nodes,
+  such as the latest height, and then access that in arbitrary
+  reactors.
```

Comment:
```suggestion
- Is it possible to store additional (potentially arbitrary)
  data into the DHT as part of the heartbeats between nodes,
  such as the latest height, and then access that in the
  reactors.
```

This is a neat idea, though I imagine we would have to be very careful about the security implications (e.g., encrypting and/or signing such data, since the DHT is not protected by the BFT resistance of consensus, so anything we put there needs some basis for trust).

Reply:
huge fan of this! it would help organise networks and peers. Most likely reduce amount hops when searching for things.

Reply:
This could relate well with the peer info work mentioned above

Reply:
I also share the same concerns as Michael in regards to the security implications of this. I would recommend starting with a small prescribed set of data yet allow for future additions. I don't think we should allow the application to have some input on this just yet but it could be something we consider in the future.

Reply:
Also what are the characteristics of DHT's. Do they work well with lots of writes. Does it make sense for writes by each node at each height?

Reply:
I was talking to someone from Celestia yesterday who had used DHT's in LibP2P and said it wasn't designed for frequent writes and believed it should just be used for peer discovery and potentially service discovery (as we don't expect nodes to change their services too frequently). For the frequent broadcasting of arbitrary data like latest height we should probably use some other mechanism (might even be a small reactor)